Index expanded dims before checking memory overlap

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #98656
* #98616

As the comment for `get_expanded_dims` says:

```
# copy_ fails when trying to write to tensors with memory overlap,
# for expanded dimensions (a dimension which used to have size 1 -> ?)
# we can select one element from that dimension and write to it
# to achieve writing to all values of that dimension of the input tensor
```

We were doing this for the copy, for not for checking if we could copy. Update it so we index then check for memory overlap. This covers all of the `complex_striding` warnings I observed in TB.

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

eellison committed:
Index expanded dims before checking memory overlap

[ghstack-poisoned]

ngimel commented on test/inductor/test_cudagraph_trees.py:
```diff
@@ -387,6 +387,21 @@ def foo(x):
+        def test_expanded_inputs(self):
+            x = torch.empty_strided([4, 512], [0, 1], device="cuda")
+            for i in range(4):
+                x[i].random_(0, 1)
```
but why?

eellison replied:
Did not see rand_strided api…

ngimel replied:
yeah but you still don't need a loop, it's an expanded tensor, so either do `x = torch.rand(1, 512, device="cuda").expand(4, 512)`, or `empty_strided` like you have and `x[0].random_(0,1)`. Anyway, it doesn't matter, just a bit confusing the way it's written.

eellison replied:
I did it this way because the logic of the PR is that we only need to access one element from the tensor, so if we did it in the test too that would have been kind of circular logic (even though we know it's correct)

ngimel replied:
But it's not really part of the test, it's preparing inputs for the test, doesn't matter how you do it as long as inputs are what you want (a tensor with the needed content and metadata)

yf225 commented on torch/_inductor/compile_fx.py:
```diff
@@ -404,25 +405,37 @@ def static_input(x):
+    ]
+
+    # copy over input values for fresh allocations
+    for idx, (x, expanded_dims) in enumerate(zip(inputs, expanded_dims)):
```
I got `UnboundLocalError: local variable 'expanded_dims' referenced before assignment` on this line, maybe we want:

for idx, (x, expanded_dims) in enumerate(zip(inputs, **inps_expanded_dims**)):

ngimel replied:
and a test that would catch it

yf225 commented:
@eellison This helps resolve the "skipping cudagraphs due to complex input striding" warning in my case. Thanks so much!

It would be super awesome to have cudagraphs overlapping check to only trigger on the non-activation inputs too (as @Chillee suggested) 😃

eellison committed:
Update on "Index expanded dims before checking memory overlap"


As the comment for `get_expanded_dims` says:

```
# copy_ fails when trying to write to tensors with memory overlap,
# for expanded dimensions (a dimension which used to have size 1 -> ?)
# we can select one element from that dimension and write to it
# to achieve writing to all values of that dimension of the input tensor
```

We were doing this for the copy, for not for checking if we could copy. Update it so we index then check for memory overlap. This covers all of the `complex_striding` warnings I observed in TB.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

eellison commented:
@pytorchbot merge
